File: 2019/day1.go
package main

import (
	"github.com/lukechampine/advent/utils"
)

var input = utils.ReadInput("day1_input.txt")
var inputInts = utils.ExtractInts(input)

func calcFuel(m int) int {
	m = (m / 3) - 2
	if m <= 0 {
		return 0
	}
	return m + calcFuel(m)
}

func main() {
	// part 1
	var sum int
	for _, m := range inputInts {
		sum += (m / 3) - 2
	}
	utils.Println(sum)

	// part 2
	sum = 0
	for _, m := range inputInts {
		sum += calcFuel(m)
	}
	utils.Println(sum)
}

File: 2019/day1.zig
const std = @import("std");
const utils = @import("utils.zig");

fn calcFuel(m: usize) usize {
    var mm = m / 3;
    return mm - std.math.min(mm, 2);
}

fn calcFuelRec(m: usize) usize {
    var mm = calcFuel(m);
    return if (mm == 0) 0 else mm + calcFuelRec(mm);
}

pub fn main() void {
    const input = utils.readFile("day1_input.txt");

    // part 1
    var it = utils.IntIterator(usize).new(input, "\n");
    var sum: usize = 0;
    while (it.next()) |m| {
        sum += calcFuel(m);
    }
    utils.println(sum);

    // part 2
    it = utils.IntIterator(usize).new(input, "\n");
    sum = 0;
    while (it.next()) |m| {
        sum += calcFuelRec(m);
    }
    utils.println(sum);
}
